### PR TITLE
Resolve xref within link

### PR DESCRIPTION
**Pull request by live1206**
As [V2](https://dotnet.github.io/docfx/tutorial/links_and_cross_references.html#different-syntax-of-cross-reference) describes, we should also support `xref` within the link.
```
[link_text](xref:uid_of_another_file)
```
This should be resolved to:
```
<a href="path_of_another_file">link_text</a>
```

### CODE REVIEW

File: docs/designs/xref.md [hunk 1]
```diff
@@ -136,6 +136,17 @@ outputs:
 ```
   - User can also define which property to display for the referenced uid
 
+### Refer to uid within markdown link
+In markdown link url, the user can also refer to uid.
+```
+[link_text](xref:uid_of_another_file)
```
Comment (928PJY) [inline docs/designs/xref.md]: what if the uid is not existed?

Comment (live1206) [reply docs/designs/xref.md]: test added for uid not found

File: docs/designs/xref.md [hunk 2]
```diff
@@ -136,6 +136,17 @@ outputs:
...
+```
+It should be resolved as:
+```
+<a href="path_of_another_file">link_text</a>
+```
+In this case, display text will be ignored unless `link_text` is empty.
```
[the remark this answers is not in the document]
Comment (live1206) [reply docs/designs/xref.md]: Updated test that displayProperty will be ignored

File: docs/specs/xref.yml [hunk 3]
```diff
@@ -321,3 +321,36 @@ outputs:
...
+    title: Title from yaml header a
+    fullName: full name
+    uid: a
+    ---
+  docs/b.md: | 
+    Link to [title](xref:a?displayProperty=fullName#bookmark)
+outputs:
+  docs/a.json:
+  docs/b.json: |
+    { "content": "<p>Link to <a href=\"/docs/a#bookmark\">title</a></p>" }
```
Comment (superyyrrzz) [inline docs/specs/xref.yml]: >/docs/a#bookmark [](start = 39, length = 16)

resolve to relative path for intermal xref?

File: src/docfx/build/dependency/Resolve.cs [hunk 4]
```diff
@@ -79,24 +81,33 @@ public static (Error error, string content, Document file) TryResolveContent(thi
...
+            var errors = new List<Error>();
             var (error, file, redirectTo, query, fragment, isSelfBookmark, _) = TryResolveFile(relativeTo, href);
+            errors.AddIfNotNull(error);
+
+            if (href.StartsWith("xref:"))
+            {
+                var (uidError, uidHref, _) = TryResolveXref(href.Substring("xref:".Length), xrefMap, dependencyMapBuilder, relativeTo);
+                errors.AddIfNotNull(uidError);
+                return (errors, uidHref, fragment, file);
+            }
```
Comment (superyyrrzz) [inline src/docfx/build/dependency/Resolve.cs]: resolve file here? `else { TryResolveFile() }`

File: docs/specs/xref.yml (hunk 3 above)
Comment (live1206) [reply docs/specs/xref.yml]: I will fix this in another PR

File: src/docfx/build/dependency/Resolve.cs (hunk 4 above)
Comment (live1206) [reply src/docfx/build/dependency/Resolve.cs]: You are right, this file is not the one we should return while resolving xref

File: src/docfx/build/dependency/Resolve.cs [hunk 5]
```diff
@@ -79,24 +81,33 @@ public static (Error error, string content, Document file) TryResolveContent(thi
...
 
-        public static (Error error, string href, string fragment, Document file) TryResolveHref(this Document relativeTo, string href, Document resultRelativeTo)
+        public static (List<Error> errors, string href, string fragment, Document file) TryResolveHref(this Document relativeTo, string href, Document resultRelativeTo, XrefMap xrefMap = null, DependencyMapBuilder dependencyMapBuilder = null)
         {
             Debug.Assert(resultRelativeTo != null);
+            var errors = new List<Error>();
+
+            if (href.StartsWith("xref:"))
+            {
+                var (uidError, uidHref, _) = TryResolveXref(href.Substring("xref:".Length), xrefMap, dependencyMapBuilder, relativeTo);
```
Comment (yufeih) [inline src/docfx/build/dependency/Resolve.cs]: is one `error` enough?

Comment (live1206) [reply src/docfx/build/dependency/Resolve.cs]: Here the only error we need to take care of is `uid-not-found`

File: src/Microsoft.Docs.Template/Microsoft.Docs.Template.csproj [hunk 6]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk.Razor">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
```
Comment (yufeih) [inline src/Microsoft.Docs.Template/Microsoft.Docs.Template.csproj]: revert this

**Pull request merged**